### ISSUE TITLE
3.0 rewrite

### ISSUE THREAD

**Issue by Ionaru** [Ionaru/easy-markdown-editor#447]
**Is your feature request related to a problem? Please describe.**
Current codebase of the editor contains a lot of legacy code, is difficult to maintain and depends on a toolchain that is no longer optimal anno 2022.

This is basically my wish list for v3.0.

**Describe the solution you'd like**
A complete rewrite of the codebase, split code over multiple files, written in TypeScript and SCSS.

**Describe alternatives you've considered**
- Keep the editor as it is, difficult to maintain and improve, I don't think this is feasible long-term.
- Slowly migrate to more modern code by rewriting parts, chance of being stuck inbetween legacy and modern code is high. Also difficult to keep the editor working like it did with the old code.

**Additional context**

- Fully ESM (https://github.com/Ionaru/easy-markdown-editor/issues/262).
- Ability to only import parts of the editor that the user needs (https://github.com/Ionaru/easy-markdown-editor/issues/167).
- CodeMirror 6 (https://github.com/Ionaru/easy-markdown-editor/issues/324).
- Ability to use plugins (https://github.com/Ionaru/easy-markdown-editor/issues/391).
- Easier event hooks.
- No bloated icon downloads (https://github.com/Ionaru/easy-markdown-editor/issues/491)
- Ability to use custom icons (raw SVG?)
- No automatic textarea detection (no magic bahaviour / side-effects).
- Typescript
- SCSS
- Modern toolchain (Rollup?)
- Updated standard look
- Standard dark mode
- Modern version of `marked` (https://github.com/Ionaru/easy-markdown-editor/issues/561)

**Progress**

You can follow progress on the https://github.com/Ionaru/easy-markdown-editor/tree/v3 branch.

**Comment by vanillajonathan** [comment]
Part of the modern toolchain could also be [Vite](https://vitejs.dev/).

    npm create vite@latest easy-markdown-editor -- --template vanilla-ts
    cd easy-markdown-editor
    npm install
    npm run dev

This makes it easy to develop with hot reload.

**Comment by Ionaru** [comment]
I agree and is something I'm considering, it does seem to be more focused on web applications and less on packages like easymde. Also it uses Rollup under the hood anyway.

> A build command that bundles your code with [Rollup](https://rollupjs.org/), pre-configured to output highly optimized static assets for production.
https://vitejs.dev/guide/

**Comment by Ionaru** [comment]
Maybe the new version of the editor should be a custom [Web Component](https://developer.mozilla.org/en-US/docs/Web/Web_Components)? The editor could be built using [Vue](https://vuejs.org/guide/extras/web-components.html) and exported to run anywhere.

**Comment by Vintagewally** [comment]
would be nice to have the possibility to toggle the spell-checker on and off while writing

**Comment by vanillajonathan** [comment]
> would be nice to have the possibility to toggle the spell-checker on and off while writing

I think the spellchecker should  be removed since browsers already come with built-in spellcheckers anyways.

**Comment by Ionaru** [comment]
> > would be nice to have the possibility to toggle the spell-checker on and off while writing
> 
> I think the spellchecker should  be removed since browsers already come with built-in spellcheckers anyways.

I can still see some use for it with for forms in languages that a browser would not be set to. But maybe off by default?

**Comment by vanillajonathan** [comment]
> I can still see some use for it with for forms in languages that a browser would not be set to. But maybe off by default?

In that case it is arguably useful. On the other hand it would arguably be nice to drop the dependency too tho which would result in less code, reduced bundle size and a smaller API.

**Comment by ThatXliner** [comment]
<img width="612" alt="image" src="https://user-images.githubusercontent.com/66848002/224441196-7316be65-0ea9-4043-8726-520cfb533368.png">

**Comment by umajho** [comment]
Hello. I tried to migrate the current codebase to rollup, and managed to pass all the tests.

Although the changes are not extensive, since this involves replacing a build system, it requires careful planning. I think my causal work is not sufficient to make a pull request.
Instead, I think it is better to put [what I have done](https://github.com/umajho/easy-markdown-editor/commit/2270ec9d527d946ba24ecc39ebbc7d5fd49a3045) here, which may provide some useful insights.

**Comment by Ionaru** [comment]
I've now pushed the https://github.com/Ionaru/easy-markdown-editor/tree/v3 branch with the current (incomplete) state of the rewrite.

**Comment by dwjohnston** [comment]
Can I recommend documentation for how to test as a first class concern? 

I'm currently facing an issue where due to our various customisations, the keyboard shortcuts have broken, so I'm currently going through and writing tests. But writing tests for this library is difficult with React Testing Library as JSDOM doesn't behave particularly well (eg https://github.com/jsdom/jsdom/issues/3002 https://github.com/hackforplay/common/issues/8 https://github.com/jsdom/jsdom/issues/1937). 

So basically, the ask is documentation that advises about how to test code that uses this library.

**Comment by velezh** [comment]
> I've now pushed the https://github.com/Ionaru/easy-markdown-editor/tree/v3 branch with the current (incomplete) state of the rewrite.

Hello @Ionaru !
You doing great work.
Could you please share any updates on it? Any status or progress of 3.0

Thanks a lot

**Comment by Viktor-Ivliev** [comment]
> > I've now pushed the https://github.com/Ionaru/easy-markdown-editor/tree/v3 branch with the current (incomplete) state of the rewrite.
> 
> Hello @Ionaru ! You doing great work. Could you please share any updates on it? Any status or progress of 3.0
> 
> Thanks a lot

It's starting to look like another abandoned library =(

**Comment by Ionaru** [comment]
Not abandoned.

**Comment by Viktor-Ivliev** [comment]
Is there any change?

**Comment by skythomp16** [comment]
I notice the branch hasn't been updated really since initial commit. If the effort isn't abandoned is there a way to get that branch updated so people can contribute?

**Comment by Ionaru** [comment]
Not abandoned, but I haven't had the free time to work on it. The current state is up-to-date.

Contributions are always welcome! Any pull request can target the 3.0 branch.

**Comment by sopoforic** [comment]
> Not abandoned, but I haven't had the free time to work on it. The current state is up-to-date.

Can you give a little information on what the current state is? Is it usable but lacking some features, broken in some major way? And if it's not currently usable, do you have a rough estimate of how much work might be needed to get it to at least a usable-as-a-beta state?

**Comment by Ionaru** [comment]
I would not recommend using it for any customer facing service. But it is usable and displays markdown pretty well.

The biggest problem I have is with the toolbar. The logic on how to interpret what the user wants when a button is clicked is tricky and I have not gotten it right in a way that I like.

For example what should happen when the button for Bold is clicked with the cursor at the start of a word, and the end or in the middle, or with a selection. 

Maybe I'm aiming too high with this, because now typing this I notice that the GitHub version not that good either haha.

**Comment by sopoforic** [comment]
> I would not recommend using it for any customer facing service. But it is usable and displays markdown pretty well.

Besides the behavior from toolbar buttons being maybe not ideal, are there any other substantial issues? I'll be implementing some of my own buttons, anyway, and using only a limited set of features from the defaults, so I'd be okay with it if it's just the kind of thing you mention with it being unclear what the bold button's behavior should be.

Basically, if there are big problems--instability, inconsistent behavior depending on browser, that sort of thing--then, yeah, it's not going to be usable yet. But if it's just a question of some behaviors being no better than 2.0, then that would be okay with me, and I could probably devote some time to working with it and contributing relevant fixes, if any, upstream. Beta releases of the 3.0 branch on npm would make it easier to integrate into a testing branch and work toward full usability.

> Maybe I'm aiming too high with this, because now typing this I notice that the GitHub version not that good either haha.

Yeah, you don't want the perfect to be the enemy of the good.

From my perspective, if 3.0 is more or less comparable to 2.0, but uses CodeMirror 6, then I'd be happy, because the biggest problem I have is that the mobile experience isn't great, and it's my understanding CM6 fixes that.

Obviously, rushing out a new version just for the sake of doing it will end up wasting time in the long run, but holding back forever isn't great either.

**Comment by Ionaru** [comment]
> Beta releases of the 3.0 branch on npm would make it easier to integrate into a testing branch and work toward full usability.

That's a good idea, I'll setup publishing of the v3 branch.

> if 3.0 is more or less comparable to 2.0

It's extremely bare-bones, lots of features and configuration options are missing and there's no preview mode yet. So I would say it's not yet comparable.

![Image](https://github.com/user-attachments/assets/1f98c13b-a91c-425e-9db7-7fb6196b89df)

**Comment by sopoforic** [comment]
Okay, yeah, preview is a major feature that I can't do without. But having v3 available so I can start testing when it's at a minimum viable point will definitely be helpful.

Thanks for taking the time to enlighten me.

**Comment by Viktor-Ivliev** [comment]
@Ionaru Congrats on the new photo ))

**Comment by Viktor-Ivliev** [comment]
Has anyone found an alternative solution with the same functionality, but one that can be updated?  This is certainly a very cool solution, but if there were a mini release here with just an update to the latest dependencies.

**Comment by Ionaru** [comment]
Just to give an update:

Progress is going well and I'm close to releasing a beta version of V3.
With Vite+ as the build system, building & testing is much easier than with any previous tool I tried.
I'm happy with the architecture and rapidly adding the missing features.

**Comment by Viktor-Ivliev** [comment]
Although I had to opt out of your solution in favor of my own custom one (purely because of my company’s policy on updating dependencies and requirements.... in particular, `marked` dependency), I’m still very grateful to you for working on this library and for your current contributions to optimous-ui.

**Comment by Ionaru** [comment]
### EasyMDE V3 Beta 1 is available! 🎉

The first beta of EasyMDE V3 is now on npm.

V3 is a ground-up rewrite. The editor core moved from CodeMirror 5 to **CodeMirror 6**, the package is **ESM-only**, and there is now an `<easy-markdown-editor>` web component alongside the class API.

[README](https://github.com/Ionaru/easy-markdown-editor/tree/v3)
[MIGRATION GUIDE](https://github.com/Ionaru/easy-markdown-editor/blob/v3/MIGRATION.md)
[CHANGELOG](https://github.com/Ionaru/easy-markdown-editor/blob/v3/CHANGELOG.md)

---

**What's new**

- CodeMirror 6 core.
- Marked 18.
- `<easy-markdown-editor>` web component!
- WIP Plugin API. The toolbar, preview and status bar are plugins themselves.
- Automatic dark mode.
- CSS variables.
- Preview sanitized by default with DOMPurify, including the output of a custom `previewRender`.
- Bundled icons (custom icons are TODO).

**What changed or is missing**

Because this is a rewrite, a lot of V2 options were renamed, removed, or aren't wired up yet. Image upload, autosave and spell checking are not yet implemented. Custom status bar items and keybinding overrides aren't supported yet. The [migration guide](https://github.com/Ionaru/easy-markdown-editor/blob/v3/MIGRATION.md) has the full list.

---

**How to get it**

```bash
npm install easymde@beta
```

```js
import { EasyMDE } from "easymde";
import "easymde/style.css";

const textarea = document.querySelector("textarea");
new EasyMDE({ element: textarea });
```

Or skip the JS entirely and use the web component, which is registered on import:

```html
<easy-markdown-editor name="comment">**Initial** content</easy-markdown-editor>
```

There is no bundled CDN build anymore, but the ESM package loads straight from any ESM CDN:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>EasyMDE V3</title>
    <link rel="stylesheet" href="https://esm.sh/easymde@beta/style.css">
</head>
<body>
<textarea></textarea>
<script type="module">
    import { EasyMDE } from "https://esm.sh/easymde@beta";
    new EasyMDE({ element: document.querySelector("textarea") });
</script>
</body>
</html>
```

---

**Feedback very welcome**

The most useful things you can tell me:

- Does the migration guide cover your V2 setup, or did you get stuck on something it doesn't mention?
- Is there something you relied on in V2 that you need back?
- Anything rough in the new CodeMirror 6 behaviour, the styling or the API?
- And of course I'd love to hear positive feedback too!

For bugs, please open a separate issue mentioning the beta version rather than replying here, so they don't get buried in this thread.

---

I plan to release more `beta` versions before going to `rc` and a stable `3.0.0` release.

**Comment by appel** [comment]
Feels like Christmas in August! Thanks @Ionaru!
